rmf_fleet_msgs/msg/RobotMode.msg

```diff
@@ -13,5 +13,9 @@ uint32 MODE_DOCKING=7
 uint32 MODE_ADAPTER_ERROR=8
 
 uint32 MODE_CLEANING=9
+uint32 MODE_PERFORMING_ACTION=10
+uint32 MODE_ACTION_COMPLETED=11
```

Comment:
Why do we need a `MODE_ACTION_COMPLETED`?
How long with the mode remain as `MODE_ACTION_COMPLETED` until it switches to `MODEL_IDLE` or something else? If we're relying on this status to infer completion, could there be a possibility to miss the state if the robot switches to the next phase of its task?

Reply:
There was a conversation on this in one of the other PRs, and I left my thoughts on it [here](https://github.com/open-rmf/rmf_demos/pull/225#discussion_r1621551205).

The truth is, we cornered ourselves with this interface ~6 years ago since we needed to design it before actions or async services were available in ROS 2. As a result we used topics, and when using topics to carry out a call/response protocol, you have no choice but to embed some way for the response to explicitly and uniquely identify which call it's associated with.

In general it would be better for us to switch to services and/or actions for commanding the simulated robots, but I don't think we should hold back this PR on such a significant architectural change. At this point it should only be slotcar using these messages, so I don't see as much need to be very strict about design quality (especially since the design quality issues massively predate this PR).

Reply:
> If we're relying on this status to infer completion, could there be a possibility to miss the state if the robot switches to the next phase of its task?

The adapter will be blocking against any new phase beginning until it sees the `ACTION_COMPLETED` mode, so this wouldn't be an issue. 

> How long with the mode remain as MODE_ACTION_COMPLETED until it switches to MODEL_IDLE or something else?

I would suggest it remains in that mode for an indefinite amount of time until a new command is received. I believe the only time the system is making decisions based on the mode value is after a fresh command has been issued, and the mode value is only checked to ensure that it matches what the command asked for. 

Admittedly there are edge cases where this could give weird results, like if an action is the last step in a task and there is no finishing task so the robot just sits idle but its mode is `MODE_ACTION_COMPLETED`. That would look odd for anyone echoing the robot states, but it wouldn't have an effect on the dashboard or anything that we expect humans to ordinarily look at. This is also limited to simulation, so the weirdness is pretty well contained.

Reply:
iiuc, this is a workaround for the design limitations in demo fleet manager and normal FA shouldn't need to use this? I think we should document this and discourage use of this mode.

Reply:
Normal fleet adapters already shouldn't be using these messages at all. 

In the `rmf_demos_fleet_adapter` we split the implementation into `fleet_adapter.py` and `fleet_manager.py` where our users are expected to use `fleet_adapter.py` as a reference or template for developing their own adapters. These messages are only being used in `fleet_manager.py` which we generally expect our users to ignore.